### PR TITLE
Add YAML to parsers

### PR DESCRIPTION
**Pull request by mohsen1**
This is work in progress.

I am trying to understand how [`src/parsers/index.js`](https://github.com/fkling/astexplorer/blob/master/src/parsers/index.js) works. What do I need to the newly added `yaml` folder?

### CODE REVIEW

**Comment by fkling** [comment]
Awesome, thank you!

> I am trying to understand how src/parsers/index.js works. What do I need to the newly added `yaml` folder. 

You shouldn't have to do anything with that file. See for example #178 that added an SQL category.

**Comment by mohsen1** [comment]
Apparently Webpack watch wasn't picking up new files. Now problem is the AST generated by this lib is not compatible. What format the AST should be in? I can write a simple transformer.

**Comment by fkling** [comment]
astexplorer expects the AST to be a plain object. The *structure* of the object can be arbitrary. What does the parser currently return?

**Comment by mohsen1** [comment]
Ok, this is funny. I used the wrong lib. I should've used `yaml-js` not `yamljs`:

here is how the AST looks like:

```js
console.log(JSON.stringify(require("yaml-js").compose(`
foo: [1,2,3,4]
bar:
    a: 1
    b: 2
    c: 3
`), null, 4));
```
<details>

```json
{
    "tag": "tag:yaml.org,2002:map",
    "value": [
        [
            {
                "tag": "tag:yaml.org,2002:str",
                "value": "foo",
                "start_mark": {
                    "line": 1,
                    "column": 0,
                    "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                    "pointer": 1
                },
                "end_mark": {
                    "line": 1,
                    "column": 3,
                    "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                    "pointer": 4
                },
                "unique_id": "node_1"
            },
            {
                "tag": "tag:yaml.org,2002:seq",
                "value": [
                    {
                        "tag": "tag:yaml.org,2002:int",
                        "value": "1",
                        "start_mark": {
                            "line": 1,
                            "column": 6,
                            "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                            "pointer": 7
                        },
                        "end_mark": {
                            "line": 1,
                            "column": 7,
                            "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                            "pointer": 8
                        },
                        "unique_id": "node_3"
                    },
                    {
                        "tag": "tag:yaml.org,2002:int",
                        "value": "2",
                        "start_mark": {
                            "line": 1,
                            "column": 8,
                            "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                            "pointer": 9
                        },
                        "end_mark": {
                            "line": 1,
                            "column": 9,
                            "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                            "pointer": 10
                        },
                        "unique_id": "node_4"
                    },
                    {
                        "tag": "tag:yaml.org,2002:int",
                        "value": "3",
                        "start_mark": {
                            "line": 1,
                            "column": 10,
                            "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                            "pointer": 11
                        },
                        "end_mark": {
                            "line": 1,
                            "column": 11,
                            "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                            "pointer": 12
                        },
                        "unique_id": "node_5"
                    },
                    {
                        "tag": "tag:yaml.org,2002:int",
                        "value": "4",
                        "start_mark": {
                            "line": 1,
                            "column": 12,
                            "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                            "pointer": 13
                        },
                        "end_mark": {
                            "line": 1,
                            "column": 13,
                            "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                            "pointer": 14
                        },
                        "unique_id": "node_6"
                    }
                ],
                "start_mark": {
                    "line": 1,
                    "column": 5,
                    "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                    "pointer": 6
                },
                "end_mark": {
                    "line": 1,
                    "column": 14,
                    "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                    "pointer": 15
                },
                "flow_style": true,
                "unique_id": "node_2"
            }
        ],
        [
            {
                "tag": "tag:yaml.org,2002:str",
                "value": "bar",
                "start_mark": {
                    "line": 2,
                    "column": 0,
                    "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                    "pointer": 16
                },
                "end_mark": {
                    "line": 2,
                    "column": 3,
                    "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                    "pointer": 19
                },
                "unique_id": "node_7"
            },
            {
                "tag": "tag:yaml.org,2002:map",
                "value": [
                    [
                        {
                            "tag": "tag:yaml.org,2002:str",
                            "value": "a",
                            "start_mark": {
                                "line": 3,
                                "column": 4,
                                "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                                "pointer": 25
                            },
                            "end_mark": {
                                "line": 3,
                                "column": 5,
                                "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                                "pointer": 26
                            },
                            "unique_id": "node_9"
                        },
                        {
                            "tag": "tag:yaml.org,2002:int",
                            "value": "1",
                            "start_mark": {
                                "line": 3,
                                "column": 7,
                                "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                                "pointer": 28
                            },
                            "end_mark": {
                                "line": 3,
                                "column": 8,
                                "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                                "pointer": 29
                            },
                            "unique_id": "node_10"
                        }
                    ],
                    [
                        {
                            "tag": "tag:yaml.org,2002:str",
                            "value": "b",
                            "start_mark": {
                                "line": 4,
                                "column": 4,
                                "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                                "pointer": 34
                            },
                            "end_mark": {
                                "line": 4,
                                "column": 5,
                                "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                                "pointer": 35
                            },
                            "unique_id": "node_11"
                        },
                        {
                            "tag": "tag:yaml.org,2002:int",
                            "value": "2",
                            "start_mark": {
                                "line": 4,
                                "column": 7,
                                "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                                "pointer": 37
                            },
                            "end_mark": {
                                "line": 4,
                                "column": 8,
                                "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                                "pointer": 38
                            },
                            "unique_id": "node_12"
                        }
                    ],
                    [
                        {
                            "tag": "tag:yaml.org,2002:str",
                            "value": "c",
                            "start_mark": {
                                "line": 5,
                                "column": 4,
                                "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                                "pointer": 43
                            },
                            "end_mark": {
                                "line": 5,
                                "column": 5,
                                "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                                "pointer": 44
                            },
                            "unique_id": "node_13"
                        },
                        {
                            "tag": "tag:yaml.org,2002:int",
                            "value": "3",
                            "start_mark": {
                                "line": 5,
                                "column": 7,
                                "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                                "pointer": 46
                            },
                            "end_mark": {
                                "line": 5,
                                "column": 8,
                                "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                                "pointer": 47
                            },
                            "unique_id": "node_14"
                        }
                    ]
                ],
                "start_mark": {
                    "line": 3,
                    "column": 4,
                    "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                    "pointer": 25
                },
                "end_mark": {
                    "line": 6,
                    "column": 0,
                    "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
                    "pointer": 48
                },
                "flow_style": false,
                "unique_id": "node_8"
            }
        ]
    ],
    "start_mark": {
        "line": 1,
        "column": 0,
        "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
        "pointer": 1
    },
    "end_mark": {
        "line": 6,
        "column": 0,
        "buffer": "\nfoo: [1,2,3,4]\nbar:\n    a: 1\n    b: 2\n    c: 3\n\u0000",
        "pointer": 48
    },
    "flow_style": false,
    "unique_id": "node_0"
}
```

</details>

**Comment by mohsen1** [comment]
Updated to yaml-js and I can now see the AST on the right side. Node highlighting is not working though.

**Comment by mohsen1** [comment]
Found a better parser. The highlighting works but autofocus doesn't

screen shot 2017-01-03 at 12 39 30 pm

**Comment by mohsen1** [comment]
@fkling do you know why autofocus is not working?

File: src/parsers/yaml/yaml-ast-parser.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+  _ignoredProperties: new Set(['parent', 'errors']),
+  locationProps: new Set(['startPosition', 'endPosition']),
+
+  nodeToRange(node) {
+    return [node.startPosition, node.endPosition];
+  },
+
+  getNodeName(node) {
+    return node.type;
```
Comment (fkling) [inline src/parsers/yaml/yaml-ast-parser.js]: Looking at the AST you posted, you probably want to process `node.tag` here to extract the last part (`map`, `str`, etc). `node.type` doesn't seem to exist.

*edit:* Oh, I just realized that you are switched the parser again. Maybe `node.kind` can be mapped to a meaningful name?

Comment (mohsen1) [inline src/parsers/yaml/yaml-ast-parser.js]: hmmm, `node.kind` is a number. I should find the enum mapping for those numbers.

Comment (mohsen1) [inline src/parsers/yaml/yaml-ast-parser.js]: There it is https://github.com/mulesoft-labs/yaml-ast-parser/blob/84445e21f61396218de5209c5221310d4e26965f/src/yamlAST.ts#L6-L13

**Comment by mohsen1** [comment]
>  I can patch your PR locally and have a look later.

Please feel free to push to my branch

**Comment by RReverser** [comment]
@mohsen1 You don't really need to duplicate the enum, as this will mean we need to manually maintain any changes should they happen upstream. Instead, you can import `Kind` from `yaml-ast-parser` when it's loaded.

**Comment by mohsen1** [comment]
@fkling I couldn't figure out the focus issue. Other than that everything is working and this can be merged.

**Comment by fkling** [comment]
I found the issue and will push a fix after merging this branch. Thank you again!

**Pull request merged**